File: app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/summary/IngredientAnalysisTagsAdapter.java
package openfoodfacts.github.scrachx.openfood.views.product.summary;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.appcompat.widget.AppCompatImageView;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.RecyclerView;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.models.entities.analysistagconfig.AnalysisTagConfig;
import openfoodfacts.github.scrachx.openfood.utils.Utils;

public class IngredientAnalysisTagsAdapter extends RecyclerView.Adapter<IngredientAnalysisTagsAdapter.ViewHolder> {
    private final WeakReference<Context> contextRef;
    private final LayoutInflater inflater;
    private final SharedPreferences prefs;
    private final List<AnalysisTagConfig> tags;
    private final List<AnalysisTagConfig> visibleTags = new ArrayList<>();
    private OnItemClickListener onClickListener;

    // data is passed into the constructor
    public IngredientAnalysisTagsAdapter(Context context, List<AnalysisTagConfig> tags) {
        contextRef = new WeakReference<>(context);
        prefs = PreferenceManager.getDefaultSharedPreferences(context);
        this.inflater = LayoutInflater.from(context);
        this.tags = tags;
        this.visibleTags.addAll(tags);

        filterVisibleTags();
    }

    // inflates the row layout from xml when needed
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.product_ingredient_analysis_tag, parent, false);
        return new ViewHolder(view);
    }

    // binds the data to the TextView in each row
    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Context context = contextRef.get();
        if (context != null) {
            AnalysisTagConfig tag = visibleTags.get(position);
            Utils.picassoBuilder(context)
                .load(tag.getIconUrl())
                .into(holder.icon);
            Drawable background = context.getResources().getDrawable(R.drawable.rounded_button);
            background.setColorFilter(Color.parseColor(tag.getColor()), android.graphics.PorterDuff.Mode.SRC_IN);
            holder.itemView.setBackground(background);

            holder.itemView.setTag(R.id.analysis_tag_config, tag);
        }
    }

    // total number of rows
    @Override
    public int getItemCount() {
        return visibleTags.size();
    }

    // stores and recycles views as they are scrolled off screen
    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        final View background;
        final AppCompatImageView icon;

        ViewHolder(View itemView) {
            super(itemView);
            itemView.setOnClickListener(this);
            background = itemView;
            icon = itemView.findViewById(R.id.icon);
        }

        @Override
        public void onClick(View view) {
            if (onClickListener != null) {
                onClickListener.onItemClick(view, getAdapterPosition());
            }
        }
    }

    public void filterVisibleTags() {
        visibleTags.clear();
        for (AnalysisTagConfig tag :
            tags) {
            if (prefs.getBoolean(tag.getType(), true)) {
                visibleTags.add(tag);
            }
        }

        notifyDataSetChanged();
    }

    // allows clicks events to be caught
    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onClickListener = onItemClickListener;
    }

    // parent activity will implement this method to respond to click events
    public interface OnItemClickListener {
        void onItemClick(View view, int position);
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/view/summary/IngredientAnalysisTagsAdapter.kt
package openfoodfacts.github.scrachx.openfood.features.product.view.summary

import android.content.Context
import android.content.SharedPreferences
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.core.content.res.ResourcesCompat
import androidx.core.graphics.BlendModeColorFilterCompat.createBlendModeColorFilterCompat
import androidx.core.graphics.BlendModeCompat
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.models.entities.analysistagconfig.AnalysisTagConfig

class IngredientAnalysisTagsAdapter(
    private val context: Context,
    private val tags: List<AnalysisTagConfig>,
    private val picasso: Picasso,
    private val sharedPreferences: SharedPreferences,
) : RecyclerView.Adapter<IngredientAnalysisTagsAdapter.IngredientAnalysisTagsViewHolder>() {

    private val visibleTags = tags.toMutableList()
    private var onClickListener: ((View, Int) -> Unit)? = null

    // inflates the row layout from xml when needed
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): IngredientAnalysisTagsViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.product_ingredient_analysis_tag, parent, false)
        return IngredientAnalysisTagsViewHolder(view)
    }

    // binds the data to the TextView in each row
    override fun onBindViewHolder(holder: IngredientAnalysisTagsViewHolder, position: Int) {
        val tag = visibleTags[position]
        picasso.load(tag.iconUrl).into(holder.icon)

        holder.background.background = ResourcesCompat.getDrawable(
            context.resources,
            R.drawable.rounded_button,
            context.theme
        )?.apply {
            colorFilter = createBlendModeColorFilterCompat(Color.parseColor(tag.color), BlendModeCompat.SRC_IN)
        }

        holder.background.setTag(R.id.analysis_tag_config, tag)
    }

    // total number of rows
    override fun getItemCount() = visibleTags.count()

    fun filterVisibleTags() {
        visibleTags.clear()

        tags.filterTo(visibleTags) {
            sharedPreferences.getBoolean(it.type, true)
        }.sortBy { it.id }

        notifyDataSetChanged()
    }

    // allows clicks events to be caught
    fun setOnItemClickListener(listener: (View, Int) -> Unit) {
        onClickListener = listener
    }

    // stores and recycles views as they are scrolled off screen
    inner class IngredientAnalysisTagsViewHolder(
        val itemView: View
    ) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        val background = itemView
        val icon: ImageView = itemView.findViewById(R.id.icon)

        override fun onClick(view: View) {
            onClickListener?.let { it(view, bindingAdapterPosition) }
        }

        init {
            itemView.setOnClickListener(this)
        }
    }

    // data is passed into the constructor
    init {
        filterVisibleTags()
    }
}
